### ISSUE TITLE
[FEATURE REQUEST] Enhance LambdaBodyLength to something like LambdaBodyExecutableStatementCount

### ISSUE THREAD

**Issue by gdinant** [checkstyle/checkstyle#8966]
Hello guys,

Related to the latest introduced functionality https://github.com/checkstyle/checkstyle/issues/8341 . The idea is great and I would definitely vouch for any check styles that favour code readability.

However in that specific scenario, that check isn't applicable to 99% of the violation we could encounter in our code base for the following reasons:

- Counts empty lines (though empty lines usually add in visibility).

```java
myLambda -> {
  action1();
  action2();
  action3();
  action4();
  action5();
  action6();
  action7();
  action8();
  action9();
  action10();
}
```

```java
myLambda -> {
  action1();

  action2();

  action3();

  action4();

  action5();

}
```

- Isn't suitable with the usage of Builder patterns

```java
myLambda -> {
  return MyObject.builder()
    .attribute1(1)
    .attribute2(2)
    .attribute3(3)
    .attribute4(4)
    .attribute5(5)
    .attribute6(6)
    .attribute7(7)
    .attribute8(8)
    .attribute9(9)
    .attribute10(10)
    .build();
}
```

Consequently, would it be possible to have such functionality but *statement* based rather than counting purely the number of lines?

Example:

/var/tmp $ cat config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
		"-//Puppy Crawl//DTD Check Configuration 1.3//EN"
		"https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checkstyle">

	<property name="charset" value="UTF-8" />
	<property name="tabWidth" value="4" />

	<module name="TreeWalker">
		<module name="LambdaBodyLength"/>
	</module>

</module>
```

```java
import java.util.function.Supplier;

import lombok.Builder;

public class Test {

	public static void main(String ... args) {

	}

	public void testBuilder() {

		Supplier<TestClass> a = () -> TestClass.builder()
			.attribute1(1)
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();

	}

	@Builder
	public static class TestClass {

		private final Integer attribute1;
		private final Integer attribute2;
		private final Integer attribute3;
		private final Integer attribute4;
		private final Integer attribute5;
		private final Integer attribute6;
		private final Integer attribute7;
		private final Integer attribute8;
		private final Integer attribute9;
		private final Integer attribute10;

	}

}
```

```
[INFO] There is 1 error reported by Checkstyle 8.37 with /Users/gabriel.dinant/Documents/repositories/leshop/order/experimentation/target/classes/config.xml ruleset.
[ERROR] src/main/java/Test.java:[14,36] (sizes) LambdaBodyLength: Lambda body length is 12 lines (max allowed is 10).
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.568 s
[INFO] Finished at: 2020-11-06T16:38:08+01:00
[INFO] ------------------------------------------------------------------------
```

Expectations:

/var/tmp $ cat config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
		"-//Puppy Crawl//DTD Check Configuration 1.3//EN"
		"https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checkstyle">

	<property name="charset" value="UTF-8" />
	<property name="tabWidth" value="4" />

	<module name="TreeWalker">
		<module name="LambdaBodyExecutableStatementCount"/>
	</module>

</module>
```

```
[INFO] --- maven-checkstyle-plugin:3.1.1:check (default) @ experimentation ---
[INFO] You have 0 Checkstyle violations.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.508 s
[INFO] Finished at: 2020-11-06T16:37:50+01:00
[INFO] ------------------------------------------------------------------------
```

 Thanks guys and keep up the good work :).

**Comment by strkkk** [comment]
@gdinant Thanks for your feedback.

This check was implemented in same way as other length checks, e.g. AnonInnerLength and MethodLength. It counts number of lines between curly braces or, in case of expression, number of lines between first and last token.

I think what you want is to extend [ExecutableStatementCount](https://checkstyle.org/config_sizes.html#ExecutableStatementCount) check with additional token LAMBDA, correct?

**Comment by gdinant** [comment]
Yes @strkkk, it's exactly what I meant.

**Comment by strkkk** [comment]
@gdinant please modify first post and describe it in way described (see example) https://checkstyle.org/report_issue.html#How_to_request_new_feature_for_existing_functionality.3F

**Comment by gdinant** [comment]
Updated my first comment. Let me know if that's clear enough.

**Comment by romani** [comment]
New check need to be created.
Original idea of such Checks to care about visual size of block (lambda, method,...). So all is countable. It helped to detect methods that are above vertical screen size, so scrolling in required to read whole code.

Why  not to increase default size in Check to match your most log builder ?
Again if you have builder bigger than screen, it is better to refactor code.

**Comment by gdinant** [comment]
@romani It is very unlikely that we have builders longer than a the actual screen size (potentially on entities, but still quite unlikely), however and I must be honest, scrolling "vertically" isn't so much of a big deal compared to scrolling horizontally. 

What I care about is keeping the code clean and of a decent complexity => maintanable. Having a lambda with heaps of actions should be avoided and a builder is just "one" single action is to understand independently of its size.

To give you another example, this should be avoided:
````java
var result = aList.stream()
	.filter(f -> {
			do this;
			do that;
			do this & that;
			more complex actions;
			and more complex actions;
			return ...;
	}).findAny();
````

Where this should be allowed:
```java
var result = aList.stream()
	.filter(excludeThis())
	.filter(excludeThat())
	.filter(excludeThisAndThat())
	.filter(excludeMoreComplicatedStuff())
	.findAny();
```

**Comment by romani** [comment]
do you expect violation on:
```
		Supplier<TestClass> a = (input) -> {
                       input++;
                       TestClass.builder()
			.attribute1(input)
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();
              }
```
if yes, so might want to continue to limit Lambda by size but with exception when Lambda does not have `{ }` or has, in other words it has single statement.

**Comment by gdinant** [comment]
@romani Your use cases are slightly different to what I'm referring to (limiting the number of statements / complexity as a whole), for me that feature request would be plenty.

Nonetheless, regarding your second example, as it is an unnecessary affectation I would definitely remove the brackets, but I doubt you can easily distinguish the two scenarios that easily. IDEs are smart enough to provide hints to such mistakes.

In other words:

Should be valid as long as LambdaExecuteStatementCount isn't violated
```java
var a = (input) -> {
	input++;
	return TestClass.builder()
			.attribute1(input)
			.attribute10(10)
			.build();
}
```

Should be rejected
```java
var a = () -> {
	var t = TestClass.builder()
			.attribute1(input)
			.attribute10(10)
			.build();
	return t;
}
```

**Comment by romani** [comment]
@gdinant , please forget about any IDE and any other Checks.

Presume All that we have is LambdaBodyExecuteStatementCount and chunk of code. And Check should put violation (please do not use any other synonym for this word) on code or not. 

> limiting the number of statements

what limit you think is good for you ?  both cases from your comment has 2 statements but looks like first is OK and second should has violation.

As we go to details, lets talk by combination of { compact compilable code + config(exact value of limit) + imaginary output of CLI}. To be absolutely clear on each detail.

**Comment by gdinant** [comment]
> what limit you think is good for you ? both cases from your comment has 2 statements but looks like first is OK and second should has violation. I am confused.

If we stick solely to the original request and providing the following `config.xml` both scenarios should be invalid.
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
		"-//Puppy Crawl//DTD Check Configuration 1.3//EN"
		"https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checkstyle">

	<module name="TreeWalker">
		<module name="LambdaBodyExecutableStatementCount">
				<property name="max" value="1"/>
		</module>
	</module>

</module>
```` 

```
[INFO] There is 1 error reported by Checkstyle 8.37 with /Users/gabriel.dinant/Documents/repositories/leshop/order/experimentation/target/classes/config.xml ruleset.
[ERROR] src/main/java/Test.java:[14,36] (sizes) LambdaBodyExecutableStatementCount: Lambda body statement count is 2 (max allowed is 1).
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.568 s
[INFO] Finished at: 2020-11-06T16:38:08+01:00
[INFO] ------------------------------------------------------------------------
``` 

The default limit of `ExecutableStatementCount` is 20, either we keep the same default to stay somewhat consistant with that check or we go with a smaller value. I'd probably override the default with something like a maximum of `5` statements, but it's a very personal opinion.

**Comment by romani** [comment]
So you think that you will be fine with:
```
		<module name="LambdaBodyExecutableStatementCount">
				<property name="max" value="5"/>
		</module>
```
```
var a = () -> { // there are 5 statements , OK
	var t = TestClass.builder()
			.attribute1()
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();
	t = TestClass.builder()
			.attribute1(input)
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();
	t = TestClass.builder()
			.attribute1(input)
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();
	t = TestClass.builder()
			.attribute1(input)
			.attribute2(2)
			.attribute3(3)
			.attribute4(4)
			.attribute5(5)
			.attribute6(6)
			.attribute7(7)
			.attribute8(8)
			.attribute9(9)
			.attribute10(10)
			.build();
	return t;
}
```

right ?

**Comment by gdinant** [comment]
Yes, that would be fine.

**Comment by romani** [comment]
I do not see a benefit from such Check in main project, but want it so much you can implement it in our sandbox/experimental project https://github.com/sevntu-checkstyle/sevntu.checkstyle 

You can use https://checkstyle.org/config_misc.html#DescendantToken for task that you have, it might do a job that you want.

**Comment by strkkk** [comment]
@romani I don't understand why extension of ExecutableStatementCount doesn't work?

```
<module name="ExecutableStatementCount">
				<property name="tokens" value="LAMBDA"/>
		</module>
```

I do not see any reason for a specific check

**Comment by romani** [comment]
Good idea, I forgot about this Check.
Do we need add this token to list of supported to this Check ?

**Comment by strkkk** [comment]
> Do we need add this token to list of supported to this Check ?

yes

**Comment by romani** [comment]
moved to https://github.com/checkstyle/checkstyle/issues/8971
